[11.0][MIG] l10n_es_aeat_mod303: Migration to 11.0

oihane opened the pull request:
Depende de:

- [x] l10n_es_aeat - #685 

cc @agaldona

Commit:
[IMP] l10n_es_aeat: Algo de PEP8 y refactorización de código
[MIG] l10n_es_aeat_mod303: Migración a v7 del módulo con las siguientes mejoras:
- Convenciones de la comunidad.
- PEP8 en la mayoría de código.
- Icono para el módulo.
- Adaptación de vistas a las posibilidades de la v7.
- Restringidas las cuentas bancarias a las de la compañía.

Commit:
[IMP] l10n_es_aeat: Algo de PEP8 y refactorización de código
[MIG] l10n_es_aeat_mod303: Migración a v7 del módulo con las siguientes mejoras:
- Convenciones de la comunidad.
- PEP8 en la mayoría de código.
- Icono para el módulo.
- Adaptación de vistas a las posibilidades de la v7.
- Restringidas las cuentas bancarias a las de la compañía.

Commit:
Launchpad automatic translations update.

Commit:
Launchpad automatic translations update.

Commit:
[IMP] l10n_es_aeat: Mejorado mecanismo de herencia para reaprovechar más código.
[IMP] l10n_es_aeat: Añadido wizard para ver el resultado de la exportación sin necesidad de tener el módulo document.
[IMP] l10n_es_aeat: Puesta transición de estados sin workflow.
[IMP] l10n_es_aeat_mod303: Adaptado al nuevo mecanismo. Se elimina workflow anterior mediante un script de migración.

Commit:
[IMP] l10n_es_aeat: Mejorado mecanismo de herencia para reaprovechar más código.
[IMP] l10n_es_aeat: Añadido wizard para ver el resultado de la exportación sin necesidad de tener el módulo document.
[IMP] l10n_es_aeat: Puesta transición de estados sin workflow.
[IMP] l10n_es_aeat_mod303: Adaptado al nuevo mecanismo. Se elimina workflow anterior mediante un script de migración.

Commit:
Launchpad automatic translations update.

Commit:
Launchpad automatic translations update.

acysos committed:
[FIX] l10n_es_aeat: compatibilidad de la función export_to_boe para el modelo 340

acysos committed:
[FIX] l10n_es_aeat: compatibilidad de la función export_to_boe para el modelo 340

Commit:
Launchpad automatic translations update.

Commit:
Launchpad automatic translations update.

Commit:
[FIX] l10n_es_aeat_mod130/mod303: import corregido para que valga también para la v8.

Commit:
[FIX] l10n_es_aeat_mod130/mod303: import corregido para que valga también para la v8.

Commit:
Launchpad automatic translations update.

Commit:
Launchpad automatic translations update.

pedrobaeza committed:
[MIG] Movidos todos los módulos a __unported__
[MIG] Eliminados módulos no migrados a v7

pedrobaeza committed:
[MIG] Movidos todos los módulos a __unported__
[MIG] Eliminados módulos no migrados a v7

pedrobaeza committed:
[MIG] l10n_es_aeat_mod303: Modelo 303 de la AEAT totalmente refactorizado

Presentación del Modelo AEAT 303
================================

* Incluye un motor genérico para obtener los totales de códigos de impuestos
  de manera parametrizada y con posibilidad de versionado por fechas
* Incluye un motor de parametrización de las exportaciones a formato BOE
  con posibilidad de versionado por fechas.
* Se incluyen las parametrizaciones de ambos motores para el 303.

pedrobaeza committed:
[MIG] l10n_es_aeat_mod303: Modelo 303 de la AEAT totalmente refactorizado

Presentación del Modelo AEAT 303
================================

* Incluye un motor genérico para obtener los totales de códigos de impuestos
  de manera parametrizada y con posibilidad de versionado por fechas
* Incluye un motor de parametrización de las exportaciones a formato BOE
  con posibilidad de versionado por fechas.
* Se incluyen las parametrizaciones de ambos motores para el 303.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Avoid errors in first initialization

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Avoid errors in first initialization

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

juanalfonsopr committed:
[FIX] l10n_es_aeat_mod303: Corregido cálculo de campo resultado_liquidacion y corregido formulario de mod303, sustituido campo casilla_46 repetido por campo casilla_69.

juanalfonsopr committed:
[FIX] l10n_es_aeat_mod303: Corregido cálculo de campo resultado_liquidacion y corregido formulario de mod303, sustituido campo casilla_46 repetido por campo casilla_69.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Añadir documentos explicativos del modelo

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Añadir documentos explicativos del modelo

pedrobaeza committed:
[IMP] l10n_es_aeat_vat_prorrate: Enfoque definitivo

======================
AEAT - Prorrata de IVA
======================

Módulo para gestionar la prorrata del IVA en las declaraciones de la AEAT (por
el momento, en el modelo 303), según el artículo 92 de la Ley 37/1992, de 28 de
diciembre, del Impuesto sobre el Valor Añadido.

Uso
===

Realizando el modelo 303, se puede seleccionar si se desea aplicar prorrata de
IVA.

Prorrata general
----------------

En el caso de seleccionar "Prorrata general", aparece un nuevo campo llamado
"Porcentaje de prorrata de IVA" que puede ser rellenado a mano.

Además, cuando se realice el asiento de regularización, la parte proporcional
de la cuota no deducida se compensará como gasto.

Se ha incluido un calculador del porcentaje de prorrata, utilizable en 2 casos:

* Al comienzo de un nuevo ejercicio, para determinar el porcentaje provisional
  a aplicar hasta el último periodo de ese ejercicio.
* En la última declaración del ejercicio, para determinar el porcentaje
  definitivo de prorrata, y sobre el que se compensará la diferencia.

Problemas conocidos / Hoja de ruta
==================================

* La prorrata especial de IVA no está contemplada aún.

pedrobaeza committed:
[IMP] l10n_es_aeat_vat_prorrate: Enfoque definitivo

======================
AEAT - Prorrata de IVA
======================

Módulo para gestionar la prorrata del IVA en las declaraciones de la AEAT (por
el momento, en el modelo 303), según el artículo 92 de la Ley 37/1992, de 28 de
diciembre, del Impuesto sobre el Valor Añadido.

Uso
===

Realizando el modelo 303, se puede seleccionar si se desea aplicar prorrata de
IVA.

Prorrata general
----------------

En el caso de seleccionar "Prorrata general", aparece un nuevo campo llamado
"Porcentaje de prorrata de IVA" que puede ser rellenado a mano.

Además, cuando se realice el asiento de regularización, la parte proporcional
de la cuota no deducida se compensará como gasto.

Se ha incluido un calculador del porcentaje de prorrata, utilizable en 2 casos:

* Al comienzo de un nuevo ejercicio, para determinar el porcentaje provisional
  a aplicar hasta el último periodo de ese ejercicio.
* En la última declaración del ejercicio, para determinar el porcentaje
  definitivo de prorrata, y sobre el que se compensará la diferencia.

Problemas conocidos / Hoja de ruta
==================================

* La prorrata especial de IVA no está contemplada aún.

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Algunos arreglos visuales

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Algunos arreglos visuales

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat/l10n_es_aeat_mod303: Crear asiento de regularización cuando se hace la declaración

* Generador del asiento de regularización con cargo a un proveedor "Agencia
  Estatal de Administración Tributaria" creado al efecto.

Configuración
=============

Para activar la creación del asiento de regularización en un modelo, hay que
poner en el modelo correspondiente el campo allow_posting a True, y establecer
en la configuración de impuestos los conceptos que se regularizarán con el
flag "to_regularize".

pedrobaeza committed:
[IMP] l10n_es_aeat/l10n_es_aeat_mod303: Crear asiento de regularización cuando se hace la declaración

* Generador del asiento de regularización con cargo a un proveedor "Agencia
  Estatal de Administración Tributaria" creado al efecto.

Configuración
=============

Para activar la creación del asiento de regularización en un modelo, hay que
poner en el modelo correspondiente el campo allow_posting a True, y establecer
en la configuración de impuestos los conceptos que se regularizarán con el
flag "to_regularize".

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Don't ask for the bank account

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Don't ask for the bank account

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

Commit:
[ADD] l10n_es_aeat_mot303, casilla 77

Commit:
[ADD] l10n_es_aeat_mot303, casilla 77

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Mejor herencia de impuestos

El anterior método utilizado, al tener una clase y un many2one genérico,
no permitía poner los datos en múltiples modelos hijos, ya que se repetían
los IDs, y cogía los datos de un modelo en otros.

Con este nuevo sistema, se sigue utilizando una tabla, pero con doble
índice: el modelo y el ID, parecido a un campo reference, pero gestionado
por código propio en el método de cálculo.

El 303 está adaptado a este nuevo sistema, y se incluye un script de
migración que conserve los datos.

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Mejor herencia de impuestos

El anterior método utilizado, al tener una clase y un many2one genérico,
no permitía poner los datos en múltiples modelos hijos, ya que se repetían
los IDs, y cogía los datos de un modelo en otros.

Con este nuevo sistema, se sigue utilizando una tabla, pero con doble
índice: el modelo y el ID, parecido a un campo reference, pero gestionado
por código propio en el método de cálculo.

El 303 está adaptado a este nuevo sistema, y se incluye un script de
migración que conserve los datos.

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Evitar problemas en campos related no readonly. Fixes #290

Cuando se tienen campos related, aunque el valor no se cambie,
Odoo lanza una escritura en el objeto relacionado cada vez que se
guarda el objeto principal. En el caso de la moneda asociada a la
compañía, hace que aquellos usuarios que no tengan permiso para
modificar la compañía (la mayoría), falle en la creación del modelo.

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Evitar problemas en campos related no readonly. Fixes #290

Cuando se tienen campos related, aunque el valor no se cambie,
Odoo lanza una escritura en el objeto relacionado cada vez que se
guarda el objeto principal. En el caso de la moneda asociada a la
compañía, hace que aquellos usuarios que no tengan permiso para
modificar la compañía (la mayoría), falle en la creación del modelo.

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Export fields [27] and [45]

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Export fields [27] and [45]

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

omar7r committed:
[IMP] l10n_es_aeat_mod303: Se añade la gestión del iva diferido en el 303

omar7r committed:
[IMP] l10n_es_aeat_mod303: Se añade la gestión del iva diferido en el 303

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[IMP] l10n_es_aeat_vat_prorrate: Include analytics + tests

pedrobaeza committed:
[IMP] l10n_es_aeat_vat_prorrate: Include analytics + tests

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

antespi committed:
[MIG] l10n_es_aeat/303: Rest of the migration

antespi committed:
[MIG] l10n_es_aeat/303: Rest of the migration

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Make modules uninstallable

pedrobaeza committed:
[MIG] Rename manifest files

pedrobaeza committed:
[MIG] Rename manifest files

pedrobaeza committed:
[IMP] l10n_es_aeat: Several warnings and imps

pedrobaeza committed:
[IMP] l10n_es_aeat: Several warnings and imps

pedrobaeza committed:
[MIG] l10n_es_aeat/mod303: Migration to 10.0

pedrobaeza committed:
[MIG] l10n_es_aeat/mod303: Migration to 10.0

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Multi-company security rule

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Multi-company security rule

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Adapt new BOE 2017 format

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Adapt new BOE 2017 format

pedrorgil committed:
[FIX] aeat_mod_303. Corrección tipos de exportación (#465)

pedrorgil committed:
[FIX] aeat_mod_303. Corrección tipos de exportación (#465)

fuentes73 committed:
[FIX] l10n_es_aeat_mod303: Añadir la cuenta bancaria en la exportación

fuentes73 committed:
[FIX] l10n_es_aeat_mod303: Añadir la cuenta bancaria en la exportación

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Show compensate flag

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Show compensate flag

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Record design link

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Record design link

pedrobaeza committed:
[FIX] l10n_es_aeat: Avoid multi-company error due to depends

Debido a que se utiliza una sola tabla para almacenar las líneas
de los impuestos de todos los modelos, hay una limitación en el ORM de Odoo
cuando se coloca el campo one2many de dichas líneas (tax_line_ids) como
dependencia en la definición del cálculo de un campo (entrada con
@api.depends), que recalcula los campos calculados de todos los modelos con el
mismo ID que el del registro en curso, lo que puede ser un problema en entornos
multi-compañía. Una solución a ello (aunque no evita el recálculo), es poner en
esos campos calculados `compute_sudo=True`.

pedrobaeza committed:
[FIX] l10n_es_aeat: Avoid multi-company error due to depends

Debido a que se utiliza una sola tabla para almacenar las líneas
de los impuestos de todos los modelos, hay una limitación en el ORM de Odoo
cuando se coloca el campo one2many de dichas líneas (tax_line_ids) como
dependencia en la definición del cálculo de un campo (entrada con
@api.depends), que recalcula los campos calculados de todos los modelos con el
mismo ID que el del registro en curso, lo que puede ser un problema en entornos
multi-compañía. Una solución a ello (aunque no evita el recálculo), es poner en
esos campos calculados `compute_sudo=True`.

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Conditional expression is reversed + 303 with conditional

pedrobaeza committed:
[FIX] l10n_es_aeat_*: Conditional expression is reversed + 303 with conditional

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Change end year in previous export format

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Change end year in previous export format

gurneyalex committed:
[REF] i18n: Updating translation terms from weblate [ci skip]

gurneyalex committed:
[REF] i18n: Updating translation terms from weblate [ci skip]

oca-transbot committed:
OCA Transbot updated translations from Transifex

oca-transbot committed:
OCA Transbot updated translations from Transifex

pedrobaeza committed:
[FIX] l10_es_aeat_mod303: Import 2017 data file

pedrobaeza committed:
[FIX] l10_es_aeat_mod303: Import 2017 data file

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Set date_end on 2014 export config

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Set date_end on 2014 export config

oihane committed:
[MIG] l10n_es_aeat_mod303: Migration to 11.0

oihane committed:
[MIG] l10n_es_aeat_mod303: Migration to 11.0

emagdalenaC2i commented on l10n_es_aeat_mod303/README.rst:
```diff
@@ -0,0 +1,72 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+   :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
```
¿Debería ser https://www.gnu.org/licenses/agpl ? Tal como se indica en https://raw.githubusercontent.com/OCA/maintainer-tools/master/template/module/README.rst

emagdalenaC2i commented on l10n_es_aeat_mod303/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# -*- encoding: utf-8 -*-
```
¿Eliminar esta línea que ya no es necesaria en Odoo 11?

emagdalenaC2i commented on l10n_es_aeat_mod303/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+# -*- coding: utf-8 -*-
```
¿Eliminar esta línea que ya no es necesaria en Odoo 11 y los datos de copyright?

emagdalenaC2i commented on l10n_es_aeat_mod303/models/__init__.py:
```diff
@@ -0,0 +1,3 @@
+# -*- encoding: utf-8 -*-
```
¿Eliminar esta línea?

emagdalenaC2i commented on l10n_es_aeat_mod303/models/mod303.py:
```diff
@@ -0,0 +1,192 @@
+# -*- coding: utf-8 -*-
```
¿Eliminar esta línea y los datos de copyright?

emagdalenaC2i commented on l10n_es_aeat_mod303/tests/test_l10n_es_aeat_mod303.py:
```diff
@@ -0,0 +1,317 @@
+# -*- coding: utf-8 -*-
```
¿Eliminar esta línea y los datos de copyright?

pedrobaeza replied:
En los archivos de código reales no hay que quitar el copyright

emagdalenaC2i replied:
ok

emagdalenaC2i commented on l10n_es_aeat_mod303/tests/__init__.py:
```diff
@@ -0,0 +1,4 @@
+# -*- coding: utf-8 -*-
```
¿Eliminar esta línea y los datos de copyrigth?

emagdalenaC2i commented on l10n_es_aeat_mod303/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+              "AvanzOSC,"
+              "Tecnativa,"
+              "Odoo Community Association (OCA)",
+    'website': "https://odoo-community.org/",
```
Cambiar website por https://github.com/OCA/l10n-spain

pedrobaeza commented:
@emagdalenaC2i en lugar de ir comentario a comentario, ponle "Start a review" y después los pasa todos juntos.

emagdalenaC2i commented on l10n_es_aeat_mod303/__manifest__.py:
```diff
@@ -0,0 +1,31 @@
+{
+    "name": "AEAT modelo 303",
+    "version": "11.0.1.0.0",
+    'category': "Accounting & Finance",
```
Cambiar categoría a "Accounting" para que coincida con las categorías establecidas en Odoo 11. Ver https://github.com/OCA/l10n-spain/pull/666/files

emagdalenaC2i reviewed (commented):
Algunos pequeños cambios de organización y formato

pedrobaeza commented:
Ya puedes hacer rebase ahora que el módulo base está mergeado.

oihane commented:
Si alguien sabe como se puede "solucionar" el error del travis, por favor que lo comparta, no lo consigo reproducir en local.

pedrobaeza commented:
Ya puedes hacer rebase

pedrobaeza commented:
Oihane, ¿podríais ampliar los tests para considerar el código no cubierto?

Otra sugerencia que vino de la lista de la localización:
```
Al calcular una declaración del modelo 303 de IVA, si la declaración sale a compensar (resultado negativo), el campo "Cuenta de contrapartida" se mantiene con la cuenta "475000 Hacienda Pública, acreedora por IVA". 
Entendemos que ésto es incorrecto y que debería usarse la cuenta "470000 Hacienda Pública, deudora por IVA".
 ```

pedrobaeza committed:
[IMP] l10n_es_aeat_mod303: Automatic compensation and return improvement

Si no estás inscrito en el REDEME (Registro de Devolución Mensual), y la declaración
sale negativa, la única posibilidad es la de compensar, por lo que no tiene sentido
que se tenga que marcar manualmente la casilla, ya que si olvida y se exporta, Hacienda
no lo admite.

Por eso, este cambio detecta automáticamente esta situación y coloca el tipo de resultado,
eliminando la casilla extra.

Se incluye script de migración para restituir correctamente los datos de anteriores
declaraciones.

También se hace obligatoria la cuenta bancaria cuando la declaración sea a devolver, ya
que es también un dato obligatorio al presentarlo a Hacienda.

cubells commented on l10n_es_aeat_mod303/__manifest__.py:
```diff
@@ -0,0 +1,30 @@
+    "license": "AGPL-3",
+    "depends": [
+        "l10n_es",
+        "l10n_es_aeat",
```
Me da error de secuencia en una base de datos limpia. No funciona si no instalo el módulo l10n_es_account_invoice_sequence, es una falta de dependencia?

pedrobaeza replied:
Uhm, esto me resulta extraño...

cubells commented on l10n_es_aeat_mod303/static/description/icon_source.svg:
```diff
@@ -0,0 +1,99 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
+<!-- Created with Inkscape (http://www.inkscape.org/) -->
+
+<svg
```
Este fichero no hace falta, se puede borrar.

pedrobaeza replied:
Sí que hace falta. Es el fuente del icono específico.

cubells reviewed (approved):
Minor changes requested.

LGTM.

Code review and tests

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Migration script

pedrobaeza committed:
[FIX] l10n_es_aeat_mod303: Migration not needed

Due to compute non stored field

JordiBForgeFlow committed:
Change counterpart account depending on the result type

JordiBForgeFlow committed:
En a casilla de importes a compensar debe constar, EN POSITIVO,
las cuotas a compensar procedentes de periodos anteriores.

JordiBForgeFlow committed:
Se elimina campo 'bank_account_id' y 'company_partner_id' por estar
duplicados con los campos 'partner_bank_id' y 'partner_id' del modelo
del que hereda l10n.es.aeat.mod303.report.

JordiBForgeFlow commented:
@oihane te he hecho el PR https://github.com/avanzosc/l10n-spain/pull/4. Revisa los conflictos, pero a mi entender deberias tomar todos los commits de mi rama como base de trabajo, porque incorporo mejoras de la 10.0 antes de tu commit de migracion, y entonces podriamos cerrar ese PR.

JordiBForgeFlow committed:
fix

JordiBForgeFlow committed:
fix

JordiBForgeFlow committed:
fix

JordiBForgeFlow committed:
fix

oihane committed:
Merge branch '11.0-mig-l10n_es_aeat_mod303' into 11.0-mig-l10n_es_aeat_mod303

oihane committed:
[11.0][mig][l10n_es_aeat_mod303] Varias correcciones y mejorasMerge pull request #4 from Eficent/11.0-mig-l10n_es_aeat_mod303

[11.0][mig][l10n_es_aeat_mod303] Varias correcciones y mejoras

cubells commented:
Fallan los tests.

Commit:
[11.0][mig][l10n_es_aeat_mod303] Modificar error en test.

JordiBForgeFlow commented:
He limpiado la historia e incorporado las ultimas mejoras del 303 de v10 en este PR:https://github.com/OCA/l10n-spain/pull/741

Cierro este.

oihane commented:
Gracias @jbeficent
